### PR TITLE
Tuple test

### PR DESCRIPTION
**Pull request by CongMa13**
Add unit tests for new tuple utilities. 

Tried to avoid touching tuple.hpp in the first commit.

### CODE REVIEW

**Comment by cgmillette** [comment]
@CongMa13 I've retargeted this work to feature branch layout-unroll. Can you rebase your branch there?

**Comment by CongMa13** [comment]
> @CongMa13 I've retargeted this work to feature branch layout-unroll. Can you rebase your branch there?

Sure. Will do.

File: test/unit/tuple_test/device/tuple.hpp [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+    }
+
+    template <typename DataT, uint32_t VecSize>
+    __device__ static inline bool isTupleTest()
+    {
+        bool err = false;
+
+        err |= std::is_tuple<int>::value;
+        err |= !std::is_tuple<std::tuple<int, float>>::value;
+        // err |= !std::is_tuple<const std::tuple<int, float>>::value;
```
Comment (CongMa13) [inline test/unit/tuple_test/device/tuple.hpp:53]: `const tuple` and `tuple &` cannot pass the test

Comment (cgmillette) [reply test/unit/tuple_test/device/tuple.hpp:53]: Probably will need to add some definitions there

Comment (CongMa13) [reply test/unit/tuple_test/device/tuple.hpp:53]: > Probably will need to add some definitions there

It makes sense if the tuple is only used in meta-programming. Const tuple or tuple ref should not be used. 

So I can change the check to `err |= std::is_tuple<const std::tuple<int, float>>::value;`

What do you think?

Comment (cgmillette) [reply test/unit/tuple_test/device/tuple.hpp:53]: Ok, I agree

File: library/include/rocwmma/internal/tuple.hpp [hunk 2]
```diff
@@ -236,7 +236,7 @@ namespace rocwmma
     constexpr static auto reverse(T&& t)
     {
         return reverse_impl(std::forward<T>(t),
-                            std::make_index_sequence<std::tuple_size<T>::value>());
+                            std::make_index_sequence<std::tuple_size< std::remove_cv_t<std::remove_reference_t<T>>>::value>());
```
Comment (cgmillette) [inline library/include/rocwmma/internal/tuple.hpp:239]: can probably do std::decay_t instead of the remove_cv_t<remove_reference>

**Pull request merged**